src/relay/pass/type_solver.cc

```diff
@@ -38,9 +39,261 @@ class TypeSolver::Reporter : public TypeReporterNode {
+    TypeNode* node = solver_->GetTypeNode(t);
+    found_ = found_ || (var_->FindRoot() == node->FindRoot());
+  }
+
+ private:
+  TypeSolver* solver_;
+  TypeNode* var_;
+  bool found_;
+};
+
+class TypeSolver::Unifier :
+    public TypeUnifierNode, public TypeFunctor<Type(const Type&, const Type&)> {
+ public:
+  explicit Unifier(TypeSolver* solver) : solver_(solver), tv_map_({}) {}
+
+  Type Unify(const Type& src, const Type& dst) final {
+    // Known limitation
+    // - handle shape pattern matching
+    Type new_src = InstantiateTypeVar(src);
+    Type new_dst = InstantiateTypeVar(dst);
+    TypeNode* lhs = solver_->GetTypeNode(new_dst);
+    TypeNode* rhs = solver_->GetTypeNode(new_src);
+
+    // do occur check so we don't create self-referencing structure
+    if (lhs->FindRoot() == rhs->FindRoot()) {
+      return lhs->resolved_type;
+    }
+    if (lhs->resolved_type.as<IncompleteTypeNode>()) {
+      CHECK(!CheckOccurs(lhs, rhs->resolved_type))
+        << "Incomplete type " << lhs->resolved_type << " occurs in "
+        << rhs->resolved_type << ", cannot unify";
+      solver_->MergeFromTo(lhs, rhs);
+      return rhs->resolved_type;
+    } else if (rhs->resolved_type.as<IncompleteTypeNode>()) {
+      CHECK(!CheckOccurs(rhs, lhs->resolved_type))
+        << "Incomplete type " << rhs->resolved_type << " occurs in "
+        << lhs->resolved_type << ", cannot unify";
+      solver_->MergeFromTo(rhs, lhs);
+      return lhs->resolved_type;
+    } else {
+      Type resolved = this->VisitType(lhs->resolved_type, rhs->resolved_type);
+      CHECK(resolved.defined())
+        << "Unable to unify parent types: "
+        << lhs->resolved_type << " and " << rhs->resolved_type;
+      TypeNode* top = solver_->GetTypeNode(resolved);
+      solver_->MergeFromTo(lhs, top);
+      solver_->MergeFromTo(rhs, top);
+      return resolved;
+    }
+  }
+
+  // Checks whether lhs (taken to be a type var) occurs in t, meaning
+  // there is a recursive equality constraint, which should be rejected.
+  // N.b.: A tautology like ?a = ?a is okay and should be checked for
+  // *before* calling this method
+  bool CheckOccurs(TypeNode *lhs, const Type &t) {
+    OccursChecker rc(solver_, lhs);
+    return rc.Check(t);
+  }
+
+  // if t is a type var, replace with an incomplete type
+  Type InstantiateTypeVar(const Type& t) {
+    auto* tvn = t.as<TypeVarNode>();
+    if (tvn == nullptr) {
+      return t;
+    }
+
+    TypeVar tv = GetRef<TypeVar>(tvn);
+    auto it = tv_map_.find(tv);
+    if (it != tv_map_.end()) {
+      return (*it).second;
+    }
+
+    IncompleteType hole = IncompleteTypeNode::make(tv->kind);
+    tv_map_.Set(tv, hole);
+    return hole;
+  }
+
+  // instantiate away all type parameters in a function type
+  FuncType InstantiateFuncType(const FuncType& ft) {
+    for (auto type_param : ft->type_params) {
+      InstantiateTypeVar(type_param);
+    }
+
+    // to avoid error when substituting type vars (TypeMutator
```

Comment:
Could we refactor this into a helper on function types which take a set of types and returns new type, this can just be implemented as generating the correct number of type vars and then calling that method.

Reply:
Could you be more specific about what the helper function should do? I'm not certain I understand. Do you mean a version of `Bind` that doesn't error out if it has to substitute a non-type var for a type var?

Reply:
I think I have an idea that I'll implement shortly, you can comment as to whether it's what you meant.

Reply:
After fiddling around with this, I've come to the conclusion that any such refactor would basically end up reimplementing `Bind()`, so I think it would be best to just use `Bind()`, unless you have a more specific way in which this should differ.

Reply:
Also, once we have generalization worked out, I don't think the TypeSolver should have to handle instantiation for type vars at all (currently we're duplicating logic: type inference for call nodes instantiate function types and so does the solver).